server: allow computed fields to have access to Hasura's session variables (fix #3846) (#4486)
hasura/graphql-engine: docs/graphql/manual/schema/computed-fields.rst

     A computed field can be added to a table using the :ref:`add_computed_field <api_computed_field>`
     metadata API

Accessing Hasura session variables in computed fields
-----------------------------------------------------

> It can be useful to have access to the session variable from the SQL function defining a computed field.  For instance, suppose we are recording which users have liked which article, using a table ``article_likes`` specifying a many-to-many relationship between ``article``s and ``user``s.  Then it may be useful to know if the current user liked the current article, and expose this as a *Boolean* computed field on ``article``.

Reviewer: ```suggestion
It can be useful to have access to the session variable from the SQL function defining a computed field.  For instance, suppose we want to record which users have liked which articles. We can do so using a table ``article_likes`` that specifies a many-to-many relationship between ``article`` and ``user``. In such a case it can be useful to know if the current user has liked a specific article, and this information can be exposed as a *Boolean* computed field on ``article``.
```

Reviewer: When we talk about the many-to-many relationship, can we link to the respective docs page: https://hasura.io/docs/1.0/graphql/manual/schema/relationships/database-modelling/many-to-many.html?

Reviewer: Also, I made the suggestion to remove the `s`s right after the table names, since it screws up the formatting: 
![image](https://user-images.githubusercontent.com/5722022/79969196-710ca580-8491-11ea-8ac9-18b10f932bf1.png)

> Use the :ref:`add_computed_field` API to add a function, and specify the name of the argument taking a session argument in ``session_argument``.  The session argument will be a JSON object where keys are session variable names (in lower case) and values are strings.  Use the ``->>`` JSON operator to fetch the value of a session variable as shown in the following example.

Reviewer: ```suggestion
Use the :ref:`add_computed_field` API to add a function, and specify the name of the argument taking a session argument in ``session_argument``.  The session argument is a JSON object where keys are session variable names (in lower case) and values are strings.  Use the ``->>`` JSON operator to fetch the value of a session variable as shown in the following example.
```

.. code-block:: plpgsql

>       -- 'hasura_session' will be session argument

Reviewer: ```suggestion
      -- 'hasura_session' will be the session argument
```

      RETURNS boolean AS $$
      SELECT EXISTS (
          SELECT 1
          FROM article_likes A
          WHERE A.user_id = hasura_session ->> 'x-hasura-user-id' AND A.article_id = article_row.id
      );
      $$ LANGUAGE sql STABLE;

.. code-block:: http

   POST /v1/query HTTP/1.1
   Content-Type: application/json
   X-Hasura-Role: admin

   {
       "type":"add_computed_field",
       "args":{
           "table":{
               "name":"article",
               "schema":"public"
           },
           "name":"liked_by_user",
           "definition":{
               "function":{
                   "name":"article_liked_by_user",
                   "schema":"public"
               },
               "table_argument":"article_row",
               "session_argument":"hasura_session"
           }
       }
   }

.. graphiql::
  :view_only:
  :query:
     query {
       article(where: {id: {_eq: 3}}) {
         id
         liked_by_user
       }
     }
  :response:
    {
        "data": {
            "article": [
                {
                    "id": "3",
                    "liked_by_user": true
                }
             ]
        }
    }

.. note::

>    The specified session argument will not be included in the ``<function-name>_args`` input object in the GraphQL schema.

Reviewer: ```suggestion
   The specified session argument is not included in the ``<function-name>_args`` input object in the GraphQL schema.
```
